### ISSUE TITLE
A potential use case for higher?

### ISSUE THREAD

**Issue by 9yte** [facebookresearch/higher#68]
Hi,

First of all, I apologize if here is not the right place to ask this question. Unfortunately, I didn't find a Q&A forum. And, I should say that this is an amazing project. 

Imagine I want to cheat in training my neural network and do the following: I want to change my training samples (only X, not the labels) in such a way that if I train one particular model on that for two epochs, some particular loss value (e.g., the loss value for predicting one particular sample from the training set) is smaller compared to the case that I train the model on the original training samples. To me, this task seems a perfect match for `higher`, yeah? And if yes, should I do that as the following?
```python
epochs = 2
meta_opt = torch.optim.Adam(X_train.parameters(), lr=meta_lr)
inner_opt = torch.optim.Adam(model.parameters(), lr=inner_lr)
with higher.innerloop_ctx(model, inner_opt, copy_initial_weights=False) as (fmodel, diffopt):
    for epoch in range(epochs):
        loss_sum = 0
        for x, y in zip(X_train, y_train):
            pred = fmodel(x)
            loss_sum += loss(pred, y)
        loss_sum = loss_sum / len(X_train)
        diffopt.step(loss_sum)

    pred = fmodel(x_important)
    important_loss = loss(pred, y_important)
    
important_loss.backward()
meta_opt.step()
meta_opt.zero_grad()
```

where `(x_important, y_important)` is the sample for which we want to have a small classification loss eventually. Assume that X_train is a list of `torch.nn.Parameter` instances, i.e., the training samples are trainable.

I really appreciate your time.

Thanks!

**Comment by egrefen** [comment]
In theory this is a use case for higher. However be aware the space complexity of running `higher` (or any inner loop gradient-based meta-learning algorithm) is `O(mn)` where `m` is the number of parameters in the model and `n` is the number of steps the model is run for before a meta-loss is taken. In your case, this `n == epochs * len(X_train)`. Can you fit this many copies of your model and its activations in memory? If so, you're welcome to try this, and if not, `higher` isn't the right way to do what you're doing (try Bayesian Optimization, or just REINFORCE, instead).

The usage I would recommend, with `higher`, here, would be to do a monte carlo estimate of your training loop as currently specified, i.e. instead of iterating through all values in `X_train`, sample at most `k`, where `k` is roughly equal to `total_memory_available / epochs`.

**Comment by egrefen** [comment]
Closing this issue as I believe this answers your question, but if there are follow up questions please just ask and reopen.
